zerver/lib/narrow.py

```diff
@@ -1389,47 +1391,68 @@ def fetch_messages(
+            if anchor is None:
+                # `anchor=None` corresponds to the anchor="first_unread" parameter.
+                anchor = find_first_unread_anchor(
+                    sa_conn,
+                    user_profile,
+                    narrow,
+                )
+
+            anchored_to_left = anchor == 0
+
+            # Set value that will be used to short circuit the after_query
+            # altogether and avoid needless conditions in the before_query.
+            anchored_to_right = anchor >= LARGER_THAN_MAX_MESSAGE_ID
+            if anchored_to_right:
+                num_after = 0
+
+            first_visible_message_id = get_first_visible_message_id(realm)
+
+            query = limit_query_to_range(
+                query=query,
+                num_before=num_before,
+                num_after=num_after,
+                anchor=anchor,
+                include_anchor=include_anchor,
+                anchored_to_left=anchored_to_left,
+                anchored_to_right=anchored_to_right,
+                id_col=inner_msg_id_col,
+                first_visible_message_id=first_visible_message_id,
             )
 
-        anchored_to_left = anchor == 0
-
-        # Set value that will be used to short circuit the after_query
-        # altogether and avoid needless conditions in the before_query.
-        anchored_to_right = anchor >= LARGER_THAN_MAX_MESSAGE_ID
-        if anchored_to_right:
-            num_after = 0
-
-        first_visible_message_id = get_first_visible_message_id(realm)
-
-        query = limit_query_to_range(
-            query=query,
-            num_before=num_before,
-            num_after=num_after,
-            anchor=anchor,
-            include_anchor=include_anchor,
-            anchored_to_left=anchored_to_left,
-            anchored_to_right=anchored_to_right,
-            id_col=inner_msg_id_col,
-            first_visible_message_id=first_visible_message_id,
-        )
+            main_query = query.subquery()
+            query = (
+                select(*main_query.c)
+                .select_from(main_query)
+                .order_by(column("message_id", Integer).asc())
+            )
 
-        main_query = query.subquery()
-        query = (
-            select(*main_query.c)
-            .select_from(main_query)
-            .order_by(column("message_id", Integer).asc())
-        )
         # This is a hack to tag the query we use for testing
         query = query.prefix_with("/* get_messages */")
         rows = list(sa_conn.execute(query).fetchall())
 
+    if is_message_ids_pre_specified:
+        # We don't need to do any post-processing in this case.
+        return FetchedMessages(
+            rows=rows,
+            found_anchor=False,
+            found_newest=False,
+            found_oldest=False,
+            history_limited=False,
+            anchor=None,
+            include_history=include_history,
+            is_search=is_search,
+        )
```

Comment:
Where are you doing the security checks that this user has access to the message IDs specified? We definitely need tests for that. It might be that #29380 is a complete answer to that, which @prakhar1144 is currently working on getting finished, but we still need a robust test suite in case refactoring breaks the logic.

Reply:
Added a comment for it there and some tests.